### PR TITLE
Add octree data structure to track highest-block-density regions

### PR DESCRIPTION
**Pull request by chikinn**

### CODE REVIEW

File: Assets/_Scripts/Game/Environment/Node.cs [hunk 1]
```diff
@@ -44,6 +47,33 @@ void Start()
...
+    }
+
+    public List<Vector3> GetExplosionTargets(int count)
+    {
+        return blockOctree.FindDensestRegions(count);
+    }
+
+    public bool ContainsPosition(Vector3 position)
+    {
+        return Vector3.Distance(position, transform.position) < transform.localScale.x / 2;
```
Comment (chikinn) [inline Assets/_Scripts/Game/Environment/Node.cs:74]: odd that claude wants to change this line of code from `transform.localScale.x` to `transform.localScale.x / 2`.  I assume we should reject it but I could use a second set of eyes on this

File: Assets/_Scripts/Game/Environment/Node.cs [hunk 2]
```diff
@@ -104,10 +134,10 @@ public NodeItem GetCrystal()
         return Crystal;
     }
 
-    public bool ContainsPosition(Vector3 position)
+/*    public bool ContainsPosition(Vector3 position)
```
Comment (chikinn) [inline Assets/_Scripts/Game/Environment/Node.cs:137]: TODO: remove any commented code

File: Assets/_Scripts/Game/Managers/NodeControlManager.cs [hunk 3]
```diff
@@ -8,6 +8,40 @@ public class NodeControlManager : Singleton<NodeControlManager>
...
+            }
+        }
+        Node targetNode = Nodes[0];
+        int targetCount = 3;
+        List<Vector3> explosionTargets = targetNode.GetExplosionTargets(targetCount);
+        Debug.Log($"Found {explosionTargets.Count} explosion targets in node {targetNode.ID}:");
+        foreach (Vector3 target in explosionTargets)
+        {
+            Debug.Log($"Target position: {target}, Block count density: {targetNode.blockOctree.GetBlockDensityAtPosition(target)}");
+        }
```
Comment (chikinn) [inline Assets/_Scripts/Game/Managers/NodeControlManager.cs:29]: this should probably be a method of `BlockOctree`, invoked not here in NCM but at the point of application, e.g. by a boid

File: Assets/_Scripts/Game/Environment/Node.cs (hunk 1 above)
Comment (chikinn) [reply Assets/_Scripts/Game/Environment/Node.cs:74]: after a moment of research I'm still not sure.  my guess is the `/ 2` shouldn't be needed.  I'm going to err on the side of not breaking the existing codebase, and remove it

File: Assets/_Scripts/Game/Managers/BlockOctree.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+        foreach (var block in node.Blocks)
+        {
+            int octant = GetOctant(node, block.transform.position);
+            AddBlockRecursive(node.Children[octant], block);
+        }
+        node.Blocks.Clear();
+    }
+
+    private int GetOctant(OctreeNode node, Vector3 position)
```
Comment (e-choness) [inline Assets/_Scripts/Game/Managers/BlockOctree.cs:99]: "This one is clever" by @KidIgarus

File: Assets/_Scripts/Game/Managers/BlockOctree.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    public double GetBlockDensityAtPosition(Vector3 position)
+    {
+        return GetBlockDensityRecursive(root, position);
+    }
+
+    private double GetBlockDensityRecursive(OctreeNode node, Vector3 position)
+    {
+        if (node.IsLeaf)
+        {
+            return node.BlockCount / Math.Pow(node.Size, 3);
```
Comment (e-choness) [inline Assets/_Scripts/Game/Managers/BlockOctree.cs:159]: @KidIgarus pointed out that the volumn aka Math.Pow(node.Size, 3) are calculated multiple times. Might as well store volumn as a property in the tree struct.

**Comment by chikinn** [comment]
@KidIgarus @e-choness thanks for taking a look already -- might be a little premature though, I'll mark it ready for review when I think it's in a good place

**Comment by chikinn** [comment]
@gradies I think this is ready for testing.  let's hook it up to boids on mon (or feel free to peek before then), I probably won't be on slack or the ship in the interim but I'll text you with some possible times to meet

if it tests alright I'll clean up the code a bit further and submit the PR to our gurus for formal feedback

**Pull request merged**